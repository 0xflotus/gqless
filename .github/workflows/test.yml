name: Test

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [12.x]

    steps:
      - uses: actions/checkout@v1
      - uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - name: Setup
        run: |
          npm install -g yarn
          yarn
          yarn bootstrap

      - name: Setup Code Climate test-reporter
        run: |
          curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
          chmod +x ./cc-test-reporter

      - name: Test
        run: |
          ./cc-test-reporter before-build
          yarn test --coverage
          ./cc-test-reporter after-build

      - name: Build
        run: yarn build:all
