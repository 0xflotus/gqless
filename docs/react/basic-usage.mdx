---
id: basic-usage
title: Basic usage
---

> #### Prerequisites
>
> Make sure you've completed [Getting Started](/docs/getting-started) first, `./gqless` refers to the folder created there

Using gqless with React is quite simple. All components that contain `gqless` data objects, should be wrapped in `graphql()` for:

- Automatic updates
- React suspense
- Query combining

## Installation

First ensure you've installed the package from NPM:

```bash
yarn add @gqless/react
# or
npm install @gqless/react
```

## Example

```jsx
import { query } from './gqless'
import { graphql } from '@gqless/react'

const App = graphql(() => {
  return <div>{query.hello}</div>
})
```

## React suspense

gqless was designed around the concept of React Suspense. Instead of components handling loading states, they should render as if the data was always there.

## Automatic updates

Selections referenced on render, will trigger components to update whenever they change.

### Example

Updating the cache when a button is pressed:

```jsx
import { query } from './gqless'
import { graphql } from '@gqless/react'

const App = graphql(() => {
  return (
    <>
      <button onClick={() => (query.hello = 'Updates')}></button>
      {query.hello}
    </>
  )
})
```

## Query combining

gqless combines component data requirements into a single query. See [custom queries (/react/custom-queries) for details on how to split them up
