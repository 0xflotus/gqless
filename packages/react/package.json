{
  "name": "@gqless/react",
  "version": "0.0.1-alpha.12",
  "main": "dist/index.js",
  "umd:main": "dist/gqlessreact.umd.production.js",
  "module": "dist/gqlessreact.es.production.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "watch": "tsdx watch",
    "build": "tsdx build",
    "prepublish": "in-publish && yarn build",
    "test": "tsdx test --env=jsdom"
  },
  "peerDependencies": {
    "gqless": "*",
    "react": ">=16"
  },
  "devDependencies": {
    "@types/react": "^16.8.17",
    "@types/react-dom": "^16.8.4",
    "react": "^16.8.6",
    "react-dom": "^16.8.6"
  },
  "gitHead": "a6601a3f9703830b097595b2eab2ed357f529feb"
}
