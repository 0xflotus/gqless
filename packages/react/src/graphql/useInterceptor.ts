import { Accessor, Interceptor, Scheduler } from 'gqless'
import { StackContext } from '../Query'

export const useInterceptor = (stack: StackContext) => {
  // Create a new Interceptor, which tracks the usage
  // of accessors
  const interceptor = new Interceptor()

  // When a new accessor is found, retrieve the
  // scheduler associated with it
  //
  // Then call Scheduler#beginQuery, with the
  // component's stack
  const schedulers = new Set<Scheduler>()
  const interceptedAccessors = new Set<Accessor>()

  interceptor.onAccessor(accessor => {
    interceptedAccessors.add(accessor)

    if (schedulers.has(accessor.scheduler)) return
    schedulers.add(accessor.scheduler)

    stack.frames.forEach(query => {
      accessor.scheduler.beginQuery(query)
    })
  })

  return {
    interceptor,
    interceptedAccessors,
    startIntercepting() {
      interceptor.start()
    },
    stopIntercepting() {
      interceptor.stop()

      // Cleanup the previous Scheduler#beginQuery
      // calls made earlier
      schedulers.forEach(scheduler => {
        for (let i = stack.frames.length - 1; i >= 0; --i) {
          const query = stack.frames[i]
          scheduler.endQuery(query)
        }
      })
    },
  }
}
