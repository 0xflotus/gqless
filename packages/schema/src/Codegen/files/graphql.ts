import { File, CORE } from '../File'
import { Codegen } from '../Codegen'

export class GraphQLFile extends File {
  constructor(private codegen: Codegen) {
    super('graphql', false)
  }

  public generate() {
    this.import(CORE, 'GraphQL', 'QueryFetcher')
    this.import('./generated', 'schema', this.codegen.schema.queryType)
    this.import('graphql/language/printer', 'print')

    return `
      ${super.generate()}

      const endpoint = ${JSON.stringify(this.codegen.options.url || '')}

      const fetchQuery: QueryFetcher = async query => {
        const response = await fetch(endpoint, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            query: print(query),
          }),
          mode: 'cors',
        })

        if (!response.ok) {
          throw new Error(\`Network error, received status code $\{response.status}\`)
        }

        const json = await response.json()

        return json
      }

      export const graphql = new GraphQL<${
        this.codegen.schema.queryType
      }>(schema.${this.codegen.schema.queryType}, fetchQuery)

      export const query = graphql.query
    `
  }
}
