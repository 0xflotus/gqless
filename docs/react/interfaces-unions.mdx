---
id: interfaces-unions
title: Interfaces & unions
---

Interfaces and unions work out of the box.

## Optimisation

Whenever you access an interface / union inside a React component, it will make the component render once for each possible type:

```jsx
import { graphql } from '@gqless/react'

const Person = graphql(({ person }) => {
  console.log(person.__typename) // => null, Student, Owner, Employee ...

  if (person.__typename === 'Student') {
    return <div>{person.class}</div>
  }

  return <div>{person.name}</div>
})
```

This is most likely unwanted, in which case you should use `ofType`:

```jsx
import { graphql } from '@gqless/react'

const Person = graphql(({ person }) => {
  console.log(person.__typename) // => null, Student

  if (ofType(person, 'Student')) {
    return <div>{person.class}</div>
  }

  return <div>{person.name}</div>
})
```
