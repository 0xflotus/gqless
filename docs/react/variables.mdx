---
id: variables
title: Variables
---

It's good practise to use variables whenever arguments change based upon state.

## Example

```jsx
import { graphql, useVariable } from '@gqless/react'

const App = graphql(() => {
  const [id, setId] = useVariable('bob')

  return <div>{query.user({ id }).name}</div>
})
```
