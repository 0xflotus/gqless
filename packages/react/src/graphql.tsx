import { Accessor, Interceptor, Query, Scheduler, Selection } from 'gqless'
import * as React from 'react'

import { useForceUpdate } from './hooks/useForceUpdate'
import { StackContext } from './Query'

export interface IQueryOptions {
  name?: string
  seperateRequest?: boolean
  /**
   * Whether or not child components can use their own queries.
   *
   * true  | child components can use their own queries
   * false | all queries will be merged into this components query
   * null  | (default) inherited with React context
   */
  allowInheritance?: boolean | null
}

export const graphql = <Props extends any>(
  component: (props: Props) => any,
  { name, allowInheritance = null, seperateRequest = false }: IQueryOptions = {}
) => {
  const query = new Query(name, false)

  const GraphQLComponent = (props: Props) => {
    const parentStack = React.useContext(StackContext)

    const stack = React.useMemo((): StackContext => {
      if (!parentStack.inheritance) return parentStack

      return {
        ...parentStack,
        inheritance:
          allowInheritance === null
            ? parentStack.inheritance
            : allowInheritance,
        frames: seperateRequest ? [query] : [...parentStack.frames, query],
      }
    }, [seperateRequest || parentStack])

    const accessorDisposers = React.useMemo(
      () => new Map<Accessor, Function>(),
      []
    )
    const accessors = React.useMemo(() => new Set<Accessor>(), [])
    const forceUpdate = useForceUpdate()

    React.useEffect(() => {
      return () => {
        accessorDisposers.forEach(dispose => dispose())
      }
    }, [])

    // Create a new Interceptor, which can tracks the usage
    // of accessors
    const interceptor = new Interceptor()

    // When a new accessor is found, retrieve the
    // scheduler associated with it
    //
    // Then call Scheduler#beginQuery, with the
    // component's stac
    const schedulers = new Set<Scheduler>()
    const interceptedAccessors = new Set<Accessor>()

    interceptor.onAccessor(accessor => {
      interceptedAccessors.add(accessor)

      if (schedulers.has(accessor.scheduler)) return
      schedulers.add(accessor.scheduler)

      stack.frames.forEach(query => {
        accessor.scheduler.beginQuery(query)
      })
    })

    try {
      // Begin intercepting accessors, located inside
      // within component's render method
      interceptor.start()
      var returnValue = component(props)
    } catch (e) {
      throw e
    } finally {
      interceptor.stop()

      // Cleanup the previous Scheduler#beginQuery
      // calls made earlier
      schedulers.forEach(scheduler => {
        for (let i = stack.frames.length - 1; i >= 0; --i) {
          const query = stack.frames[i]
          scheduler.endQuery(query)
        }
      })
    }

    // Find all the new accessors and add to Set
    interceptedAccessors.forEach(accessor => {
      if (accessors.has(accessor)) return

      accessors.add(accessor)
      accessorDisposers.set(
        accessor,
        // Make component update when data changes
        accessor.onDataUpdate(() => {
          forceUpdate()
        })
      )
    })

    const fetchingSelections = new Set<Selection>()

    accessors.forEach(accessor => {
      // Locate accessors currently being fetched,
      // and add to Set
      if (interceptedAccessors.has(accessor)) {
        if (accessor.selection.isFetching) {
          fetchingSelections.add(accessor.selection)
        }

        return
      }

      // Remove previously used accessors, that
      // aren't required anymore
      const dispose = accessorDisposers.get(accessor)
      if (dispose) {
        accessorDisposers.delete(accessor)
        dispose()
      }
      accessors.delete(accessor)
    })

    returnValue = (
      <StackContext.Provider value={stack}>{returnValue}</StackContext.Provider>
    )

    // React suspense support
    if (fetchingSelections.size) {
      // Promise that resolves when all selections are fetched
      let resolve: Function
      let resolved = false
      const promise = new Promise(r => (resolve = r))

      fetchingSelections.forEach(selection => {
        selection.onFetched.then(() => {
          fetchingSelections.delete(selection)

          if (!fetchingSelections.size) {
            resolved = true
            resolve()
          }
        })
      })

      // We can't directly throw the promise, otherwise
      // child components (with data requirements) won't
      // render - meaning multiple requests
      //
      // To prevent this we instead return a Fragment,
      // which contains a component that throws the Promise.
      const SuspendComponent = () => {
        // This is necessary to prevent an infinite loop
        if (resolved) return null

        throw promise
      }

      return (
        <>
          {returnValue}
          <SuspendComponent />
        </>
      )
    }

    return returnValue
  }

  GraphQLComponent.displayName = (component as any).displayName
  GraphQLComponent.query = query

  return GraphQLComponent
}
