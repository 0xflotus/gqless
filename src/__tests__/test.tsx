import { render } from 'react-testing-library'
import * as React from 'react'
import { print } from 'graphql'
import { useTestQuery, queryComponent, executeQuery } from '../../test'

test('adds 1 + 2 to equal 3', async done => {
  let queries: string[] = []
  const Component = queryComponent(
    () => {
      const query = useTestQuery({
        fetchQuery(query) {
          queries.push(print(query))

          return executeQuery(query)
        },
      })

      if (!query) return null

      return (
        <div>
          {'' + query.data.user.name}
          {'' + query.data.user.age}
        </div>
      )
    },
    () => {
      expect(queries).toMatchSnapshot()
      done()
    }
  )

  render(
    <React.Suspense fallback="loading">
      <Component />
    </React.Suspense>
  )
})
