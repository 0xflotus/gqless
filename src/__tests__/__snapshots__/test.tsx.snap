// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds 1 + 2 to equal 3 1`] = `
Array [
  "query TEST {
  user {
    name
    __typename
  }
}
",
  "query TEST {
  user {
    age
    __typename
  }
}
",
]
`;
